python/openassetio/pluginSystem/PluginSystem.py

```diff
@@ -29,56 +36,96 @@ class PluginSystem(object):
-                        self.__logger.kDebug)
-                    continue
+        self.__logger.log("PluginSystem: Searching %s" % paths, self.__logger.kDebug)
 
-                # Make a unique namespace to ensure the plugin identifier is all that
-                # really matters
-                moduleName = hashlib.md5(bundlePath.encode("utf-8")).hexdigest()
+        for path in paths.split(os.pathsep):
 
-                try:
+            if not os.path.isdir(path):
+                msg = "PluginSystem: Skipping as it is not a directory %s" % path
+                self.__logger.log(msg, self.__logger.kDebug)
+
+            for item in os.listdir(path):
+
+                itemPath = os.path.join(path, item)
+
+                if os.path.isdir(itemPath):
+                    # The directory could be a package, check for __init__.py
+                    initFile = os.path.join(itemPath, "__init__.py")
+                    if os.path.exists(initFile):
+                        itemPath = initFile
+                    else:
+                        msg = "PluginSystem: Ignoring as it is not a python package " \
+                              "contianing __init__.py %s" % itemPath
+                        self.__logger.log(msg, self.__logger.kDebug)
+                        continue
+                else:
+                    # Its a file, check if it is a .py/.pyc module
+                    _, ext = os.path.splitext(itemPath)
+                    if ext not in self.__validModuleExtensions:
+                        msg = "PluginSystem: Ignoring as its not a python module %s" % itemPath
+                        self.__logger.log(msg, self.__logger.kDebug)
+                        continue
 
-                    module = imp.load_module(
-                        moduleName, None, bundlePath, ("", "", imp.PKG_DIRECTORY))
-                    if hasattr(module, 'plugin'):
-                        self.register(module.plugin, bundlePath)
+                self.__logger.log(
+                    "PluginSystem: Attempting to load %s" % itemPath,
+                    self.__logger.kDebug)
 
-                except Exception as e:
-                    msg = "PluginSystem: Caught exception loading plug-in from '%s':\n%s" % (
-                        bundlePath, e)
-                    self.__logger.log(msg, self.__logger.kError)
+                self.__load(itemPath)
 
     def identifiers(self):
+        """
+        Returns the identifiers known to the plugin system.
+
+        If @ref scan has not been called, then this will be empty.
+
+        @return `List[str]`
+        """
         return list(self.__map.keys())
 
     def plugin(self, identifier):
+        """
+        Retrieves the plugin that provides the given identifier.
+
+        @return @ref openassetio.pluginSystem.PluginSystemPlugin
+        "PluginSystemPlugin"
+
+        @exception openassetio.exceptions.PluginError Raised if no
```

Comment:
[question] The `openassetio.` prefix isn't technically necessary - I removed them in #98, should we include it?

Reply:
We should figure this out more generally. Shall we wait to see when we do the `c++`, as the namespace may well change?